Fuse ops in eager cosine_similarity while keeping the stability and the gradients

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #104772
* __->__ #104771

There was a regression in https://github.com/pytorch/pytorch/pull/31378
which was reported in https://github.com/pytorch/pytorch/issues/104564.
This PR should keep the efficiency and memory usage from the original
implementation, while keeping the stability of the latter.

This solution was already discussed in https://github.com/pytorch/pytorch/pull/31378,
but it was discarded because it modified the vector_norm in-place. The
only magic ingredient that was missing for that solution to work was to
add a `clone()` after calling the `vector_norm`.

I hope this PR takes shorter to land than https://github.com/pytorch/pytorch/issues/104564.

Fixes https://github.com/pytorch/pytorch/issues/104564

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

lezcano committed:
Fuse ops in eager cosine_similarity while keeping the stability and the gradients

There was a regression in https://github.com/pytorch/pytorch/pull/31378
which was reported in https://github.com/pytorch/pytorch/issues/104564.
This PR should keep the efficiency and memory usage from the original
implementation, while keeping the stability of the latter.

This solution was already discussed in https://github.com/pytorch/pytorch/pull/31378,
but it was discarded because it modified the vector_norm in-place. The
only magic ingredient that was missing for that solution to work was to
add a `clone()` after calling the `vector_norm`.

I hope this PR takes shorter to land than https://github.com/pytorch/pytorch/issues/104564.

Fixes https://github.com/pytorch/pytorch/issues/104564

[ghstack-poisoned]

lezcano commented:
This implements on eager the solution discussed in https://github.com/pytorch/pytorch/issues/104564#issuecomment-1625348908

albanD commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -271,71 +272,31 @@ Tensor _pdist_backward(const Tensor& grad, const Tensor& self, const double p, c
-  auto x1 = x1_.to(commonDtype).expand_symint(common_size);
-  auto x2 = x2_.to(commonDtype).expand_symint(common_size);
+  // We accept integral types but vector_norm does not
+  auto x1_int = isIntegralType(x1_.scalar_type());
```
nit: `x1_is_int`

lezcano replied:
done

albanD commented on aten/src/ATen/native/Distance.cpp:
```diff
@@ -271,71 +272,31 @@ Tensor _pdist_backward(const Tensor& grad, const Tensor& self, const double p, c
-   *
-   * The current implementation is an improvement over the previous version.
-   *
-   * Previous implementation:
```
I think the comparison with the old implementation is still relevant in terms of numerical precision.
Also this comment is still valid for the updated implementation right?

lezcano replied:
reverted

lezcano committed:
Update on "Fuse ops in eager cosine_similarity while keeping the stability and the gradients"


There was a regression in https://github.com/pytorch/pytorch/pull/31378
which was reported in https://github.com/pytorch/pytorch/issues/104564.
This PR should keep the efficiency and memory usage from the original
implementation, while keeping the stability of the latter.

This solution was already discussed in https://github.com/pytorch/pytorch/pull/31378,
but it was discarded because it modified the vector_norm in-place. The
only magic ingredient that was missing for that solution to work was to
add a `clone()` after calling the `vector_norm`.

I hope this PR takes shorter to land than https://github.com/pytorch/pytorch/issues/104564.

Fixes https://github.com/pytorch/pytorch/issues/104564

[ghstack-poisoned]

lezcano commented:
@albanD addressed the comments

albanD reviewed (approved):
SGTM

lezcano committed:
Update on "Fuse ops in eager cosine_similarity while keeping the stability and the gradients"


There was a regression in https://github.com/pytorch/pytorch/pull/31378
which was reported in https://github.com/pytorch/pytorch/issues/104564.
This PR should keep the efficiency and memory usage from the original
implementation, while keeping the stability of the latter.

This solution was already discussed in https://github.com/pytorch/pytorch/pull/31378,
but it was discarded because it modified the vector_norm in-place. The
only magic ingredient that was missing for that solution to work was to
add a `clone()` after calling the `vector_norm`.

I hope this PR takes shorter to land than https://github.com/pytorch/pytorch/issues/104564.

Fixes https://github.com/pytorch/pytorch/issues/104564

[ghstack-poisoned]
